### ISSUE TITLE
how to handle nested config parameters

### ISSUE THREAD

**Issue by wrashid** [LBNL-UCB-STI/beam#316]
how to avoid providing dummy parameters?

**Comment by carloscaldas** [comment]
# Introduction to the problem

Currently, the BEAM project uses the `tscfg` command line tool. This tool takes a configuration specification parseable by Typesafe Config and generates all the boiler-plate to make the definitions available in type safe.

This tool allow the following approach of development:

1-Develop a configuration template
2-Run the tool to generate the POJO classes that are able to read a configuration file

After the application starts the POJO try to read the configuration file.

# Limitation

The `tfcfg` has a limitation related handle complex objects and default values. For example:

Given the following configuration file and template

**Template**
```
endpoint {
  url: "String | http://example.net"
}
notRelatedProperty: "String"
```

**Valid configuration file 1**: no need to use the default value
```
endpoint {
  url: "http://google.com"
}
notRelatedProperty: "value1"
```

**Valid configuration file 2**: in this case the default value will be used
```
endpoint {
}
notRelatedProperty: "someValue"
```

**Invalid configuratin file**: the POJO is not able to create the complex object and fill the default value
```
notRelatedProperty: "someValue"
```

**Comment by JustinPihony** [comment]
See https://github.com/carueda/tscfg/issues/33 for updates. Including the PR I just proposed - https://github.com/carueda/tscfg/pull/34

**Comment by carueda** [comment]
Thanks again @JustinPihony for your contribution to `tscfg` to overcome its previous limitation about absent optional objects. I just made a release (0.9.4) of the new command-line program and triggered the publication to maven central (some folks are using the direct fat jar from there).  But I'll also taking the liberty of tagging @timvlaer here (hope that's ok), who has been maintaining the maven plugin (https://github.com/timvlaer/tscfg-maven-plugin) so he's aware of this sooner than later ; ) and you guys could consider continuing using `tscfg` for your project.

**Comment by timvlaer** [comment]
Thanks @carueda for the heads up. I just released 0.4.0 of the [tscfg-maven-plugin](https://github.com/timvlaer/tscfg-maven-plugin) with your update included.
